R/eval-walk.R

```diff
@@ -110,11 +110,22 @@ ensure_named <- function(pos,
   check_empty(pos, allow_empty, error_arg, call = call)
 
   if (!allow_rename && any(names2(pos) != "")) {
-    cli::cli_abort(
-      "Can't rename variables in this context.",
-      class = "tidyselect:::error_disallowed_rename",
-      call = call
-    )
+    if (is.null(error_arg)) {
+      cli::cli_abort(
+        "Can't rename variables in this context.",
+        class = "tidyselect:::error_disallowed_rename",
+        call = call
+      )
+    } else {
+      cli::cli_abort(c(
+        "Can't rename variables in this context.",
+        i = "{.arg {error_arg}} can't be renamed."
+        ),
+        class = "tidyselect:::error_disallowed_rename",
+        call = call
+      )
+    }
```

Comment:
In https://github.com/r-lib/tidyselect/issues/336 I think I was hoping that we could somehow mention the expression where the rename happens in the error message.

`ensure_named()` is only called by `vars_select_eval()` and that does have the `expr`. So you could pass `expr` through to `ensure_named()` and use `as_label(expr)` on it to get something nice to include in the error message. Like

```r
      expr <- as_label(expr)
      cli::cli_abort(c(
        "Can't rename variables during selection.",
        i = "Rename detected in {.arg {error_arg}}, with expression {.code {expr}}."
        ),
        class = "tidyselect:::error_disallowed_rename",
        call = call
      )
```

For something like `c(all_of(x), y)` if the rename happens in the `all_of(x)`, it would still report the full `c(all_of(x), y)` as the place where the rename happened, but I think that is ok, it is hard to be more precise. I still think this would be a big improvement.


---

I think adding the argument name is nice, but maybe not quite as good as it could be on its own. For example this dplyr error from https://github.com/tidyverse/dplyr/issues/6745 would have been:

```r
#> Error in `rename_with()`:
#> ! Can't rename variables in this context.
#> i `.cols` can't be renamed.
```

But the user typed in `mtcars %>% rename_with(toupper, all_of(v))` so `.cols` isn't immediately meaningful to them. But this:

```r
#> Error in `rename_with()`:
#> ! Can't rename variables in this context.
#> i Rename detected in `.cols`, with expression `all_of(v)`.
```

is probably enough for them to figure out where to look next, because it refers to something in their code

WDYT @lionel- ?

Reply:
I like it!

Reply:
I think we could use the "In argument" pattern for consistency.

```r
#> Error in `rename_with()`:
#> ! Can't rename variables in this tidyselect context.
#> i In argument: `.cols = all_of(v)`.
```